enhancements/installer/azure-support-known-cloud-environments.md

```diff
@@ -0,0 +1,231 @@
+
+1. The Machines / MachineSets are created by the installer or users once but new objects would require the user to specify these values again.
+
+2. The requirement for clouds per Machine doesn't seem to provide major value.
+
+### Risks and Mitigations
+
+1. Picking correct instance type defaults for various clouds. The user UX when the installer fails to find default instance types for the a cloud is not great as it just errors out, but I think a little higher bar for non-public Azure cloud environments should not be too hard.
+
+### Test Plan
+
+TODO
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, or as something else. Initial proposal
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this
+enhancement:
+- Maturity levels - `Dev Preview`, `Tech Preview`, `GA`
+- Deprecation
+
+Clearly define what graduation means.
+
+#### Examples
+
+These are generalized examples to consider, in addition to the aforementioned
+[maturity levels][maturity-levels].
+
+##### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+
+##### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+
+**For non-optional features moving to GA, the graduation criteria must include
+end to end tests.**
+
+### Upgrade / Downgrade Strategy
+
+The changes to the API objects are backwards compatible and therefore there shouldn't be any specific concerns w.r.t to upgrades or downgrades.
+
+### Version Skew Strategy
+
+The changes to the API objects are backwards compatible and therefore there shouldn't be any specific concerns w.r.t to upgrades or downgrades.
+
+## Implementation History
+
+None
+
+## Alternatives
+
+We could treat all the non-public Azure cloud environments as Custom Azure cloud environment requiring the users to provide resource management endpoint. But that would not be great UX because,
+
+- the Azure SDKs support these known environments natively and it will be a lot more easy for users to say `UsGovernmentCloud` instead of knowing the URL.
+- Extra validations are required for URLs provided.
+- The installer can control the known cloud names easily.
+- The installer users the azurerm terraform provided that doesn't have support of custom resource endpoint URLs and would require use of the azurestack terraform provider. A change like that would be a lot more involved and therefore suporting known cloud names is a lot more easier to satisfy the goals.
```

Comment:
For reference, there's an issue (still open) about custom clouds support on the Terraform provider: https://github.com/terraform-providers/terraform-provider-azurerm/issues/6816